### ISSUE TITLE
[core] Video stuttering with `SUPPORT_PARTIALBUSY_WAIT_LOOP` on Linux

### ISSUE THREAD

**Issue by flashback-fx** [raysan5/raylib#2434]
### Issue description

There seems to be a regression in the rcore.c WaitTime function that causes video stuttering when compiling with SUPPORT_PARTIALBUSY_WAIT_LOOP for PLATFORM_DESKTOP running under Linux (I am not sure if other platforms are affected). I have been testing different raylib version tags and this stuttering does not happen in 3.7.0, and happens in 4.0.0. The cause seems to be changes made to the WaitTime function of rcore.c (previously Wait in core.c).

This is the output of doing
`git diff 3.7.0 master -- core.c rcore.c`

```
-static void Wait(float ms)
+void WaitTime(float ms)
 {
-#if defined(PLATFORM_UWP)
-    UWPGetSleepFunc()(ms/1000);
-    return;
-#endif
-
 #if defined(SUPPORT_BUSY_WAIT_LOOP)
-    double prevTime = GetTime();
-    double nextTime = 0.0;
+    double previousTime = GetTime();
+    double currentTime = 0.0;

     // Busy wait loop
-    while ((nextTime - prevTime) < ms/1000.0f) nextTime = GetTime();
+    while ((currentTime - previousTime) < ms/1000.0f) currentTime = GetTime();
 #else
-    #if defined(SUPPORT_HALFBUSY_WAIT_LOOP)
-        #define MAX_HALFBUSY_WAIT_TIME  4
-        double destTime = GetTime() + ms/1000;
-        if (ms > MAX_HALFBUSY_WAIT_TIME) ms -= MAX_HALFBUSY_WAIT_TIME;
+    #if defined(SUPPORT_PARTIALBUSY_WAIT_LOOP)
+        double busyWait = ms*0.05;     // NOTE: We are using a busy wait of 5% of the time
+        ms -= (float)busyWait;
     #endif

+    // System halt functions
     #if defined(_WIN32)
         Sleep((unsigned int)ms);
     #endif
@@ -4571,14 +4833,80 @@ static void Wait(float ms)
         usleep(ms*1000.0f);
     #endif

-    #if defined(SUPPORT_HALFBUSY_WAIT_LOOP)
-        while (GetTime() < destTime) { }
+    #if defined(SUPPORT_PARTIALBUSY_WAIT_LOOP)
+        double previousTime = GetTime();
+        double currentTime = 0.0;
+
+        // Partial busy wait loop (only a fraction of the total wait time)
+        while ((currentTime - previousTime) < busyWait/1000.0f) currentTime = GetTime();
     #endif
 #endif
 }
```

I've tried changing the body of the WaitTime function to one like it was in the raylib 3.7.0 core.c Wait function, but using the function and macro names that raylib 4.0.0 uses. This seems to solve the problem. This is how WaitTime looks after reverting its behavior to the one in raylib 3.7.0:

```
void WaitTime(float ms)
{
#if defined(SUPPORT_BUSY_WAIT_LOOP)
    double prevTime = GetTime();
    double nextTime = 0.0;

    // Busy wait loop
    while ((nextTime - prevTime) < ms/1000.0f) nextTime = GetTime();

#else
    #if defined(SUPPORT_PARTIALBUSY_WAIT_LOOP)
        #define MAX_PARTIALBUSY_WAIT_TIME  4
        double destTime = GetTime() + ms/1000;
        if (ms > MAX_PARTIALBUSY_WAIT_TIME) ms -= MAX_PARTIALBUSY_WAIT_TIME;
    #endif

    #if defined(_WIN32)
        Sleep((unsigned int)ms);
    #endif
    #if defined(__linux__) || defined(__FreeBSD__) || defined(__EMSCRIPTEN__)
        struct timespec req = { 0 };
        time_t sec = (int)(ms/1000.0f);
        ms -= (sec*1000);
        req.tv_sec = sec;
        req.tv_nsec = ms*1000000L;

        // NOTE: Use nanosleep() on Unix platforms... usleep() it's deprecated.
        while (nanosleep(&req, &req) == -1) continue;
    #endif
    #if defined(__APPLE__)
        usleep(ms*1000.0f);
    #endif

    #if defined(SUPPORT_PARTIALBUSY_WAIT_LOOP)
        while (GetTime() < destTime) { }
    #endif
#endif
}
```

I used an application with moving graphics and an FPS counter made in raylib to test the behavior of each version. This application can be started with VSYNC on or off and I can also cap the framerate to 60 FPS or unlock it. The testing has been made using a monitor set to a 60 FPS refresh rate. This is what I've observed:

4.0 and 4.1-dev (master):

- 60FPS VSYNC ON: Stutters (every second approximately), FPS counter oscillates between 59 and 60
- UNLOCKED FPS VSYNC ON: No stutter, FPS counter oscillates between 59 and 60
- 60FPS VSYNC OFF: Tearing, FPS counter between 59 and 60

3.7 and 4.1 with fix:

- 60 FPS VSYNC ON: Very occasional stutter, FPS counter at 60
- UNLOCKED FPS VSYNC ON: No stutter, FPS counter between 59 and 60
- 60 FPS VSYNC OFF: Slow tearing, FPS counter at 60

Also, 4.0 and 4.1-dev (master) compiled with SUPPORT_BUSY_WAIT_LOOP seem to behave like 3.7 and 4.1 with fix compiled with SUPPORT_PARTIALBUSY_WAIT_LOOP.

I am opening an issue instead of doing a pull request because I am not sure if reverting the behavior of WaitTime to the one raylib had in 3.7.0 has deeper implications (one thing I noticed it that the old method consumes more CPU). In any case, I can make a pull request if you think it can be useful for saving some time on testing.

### Environment

All that is explained here has been observed and tested in a machine running Kubuntu 20.04 (KDE Plasma desktop) with the desktop compositor disabled.
This is the information raylib reports for this machine on start up:
```
INFO: GLAD: OpenGL extensions loaded successfully
INFO: GL: Supported extensions count: 223
INFO: GL: OpenGL device information:
INFO:     > Vendor:   AMD
INFO:     > Renderer: AMD KAVERI (DRM 2.50.0, 5.4.0-107-generic, LLVM 12.0.0)
INFO:     > Version:  4.5 (Core Profile) Mesa 21.2.6
INFO:     > GLSL:     4.50
INFO: GL: DXT compressed textures supported
INFO: GL: ETC2/EAC compressed textures supported
```

Please let me know if you need more information.

### Code Example

I have uploaded a repo to my a account with a small program that uses the simplest and most effective way I found to showcase the issue. What it does is flash a color every other frame at 60FPS, like a strobing light. That means that it may pose an epilepsy risk, so please, if you really want to use it do it with caution.

Thanks a lot.

**Comment by raysan5** [comment]
@flashback-fx thanks for reporting. I reviewed the fix and it actually turns busyWait into 4 ms instead of using only 5% of target wait time; that's the reason it consumes more CPU but avoids stuttering, because it wastes more time actually in a busy wait loop.

Could you try increasing wait time to 20% for current implementation:
```c
#if defined(SUPPORT_PARTIALBUSY_WAIT_LOOP)
        double busyWait = ms*0.2;     // NOTE: We are using a busy wait of 20% of the time
        ms -= (float)busyWait;
#endif
```

Issue could also be related to the functions calls cost... Actually, I don't know why previous implementation was changed... but I imagine there was some reason...

**Comment by raysan5** [comment]
No answer from user. I will assume it worked.

**Comment by flashback-fx** [comment]
Hello,

Sorry for not answering before, I have been very busy and did not find all the time I wanted to dedicate to this issue until now.

I experimented with changes similar to what you suggested and it did not fix the issue.

I tried to change the current WaitTime code to use busy wait times of 20%, 40% and of a fixed 4ms and the resulting behavior was the same as the one I describe in my first message for versions 4.0 and 4.1-dev.

Fortunately, I found the cause of the regression. I've created a pull request with a fix (https://github.com/raysan5/raylib/pull/2503), you will find more details in the pull request description.

Thank you.
